Framework/API/src/SpectrumInfo.cpp

```diff
@@ -176,8 +176,15 @@ const Geometry::IDetector &SpectrumInfo::getDetector(const size_t index) const {
     // Else need to construct a DetectorGroup and use that
     std::vector<boost::shared_ptr<const Geometry::IDetector>> det_ptrs;
     for (const auto &id : dets) {
-      const auto detIndex = m_detectorInfo.indexOf(id);
-      det_ptrs.push_back(m_detectorInfo.getDetectorPtr(detIndex));
+      try {
+        const auto detIndex = m_detectorInfo.indexOf(id);
+        det_ptrs.push_back(m_detectorInfo.getDetectorPtr(detIndex));
+      } catch (std::out_of_range) {
+        // Workspaces can contain invalid detector IDs. Those IDs will be
+        // silently ignored here until this is fixed. Some valid IDs will exist
+        // if hasDetectors or hasUniqueDetectors has returned true, but there
+        // could still be invalid IDs.
+      }
```

Comment:
Yes, the root cause needs fixing. From the comments in `SpectrumInfo::hasDetectors()` and `SpectrumInfo::hasUniqueDetectors()` I was expecting some of the Instrument 2.0 changes will prevent workspaces having invalid detector IDs, so the for loops in those methods will then be removed, as would the try catch.

So should we fix the loaders now, and omit the try/catch, or can we do something more fundamental to prevent the possibility of having invalid detector IDs in the workspace now?